{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "97e8358d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:ToolkitGenerator:Range of slot 'treated by' (treatment) does not line with the domain of its inverse (treats)\n",
      "WARNING:ToolkitGenerator:Range of slot 'approved for treatment by' (treatment) does not line with the domain of its inverse (approved to treat)\n",
      "WARNING:ToolkitGenerator:Prefix case mismatch - supplied: NCBIGene expected: NCBIGENE\n",
      "WARNING:ToolkitGenerator:Prefix case mismatch - supplied: NCBIGene expected: NCBIGENE\n"
     ]
    }
   ],
   "source": [
    "import bmt\n",
    "import requests\n",
    "import json\n",
    "from pprint import pprint\n",
    "import csv\n",
    "\n",
    "tk = bmt.Toolkit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d50a62b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "related to\n",
      "narrow match\n",
      "caused by\n",
      "affects\n",
      "affects secretion of\n",
      "increases secretion of\n",
      "decreases secretion of\n",
      "entity regulates entity\n",
      "entity positively regulates entity\n",
      "entity negatively regulates entity\n",
      "ameliorates\n",
      "treats\n",
      "approved to treat\n",
      "exacerbates\n",
      "affects response to\n",
      "decreases response to\n",
      "increases response to\n",
      "affects metabolic processing of\n",
      "increases metabolic processing of\n",
      "decreases metabolic processing of\n",
      "affects mutation rate of\n",
      "increases mutation rate of\n",
      "decreases mutation rate of\n",
      "affects splicing of\n",
      "decreases splicing of\n",
      "increases splicing of\n",
      "affects folding of\n",
      "increases folding of\n",
      "decreases folding of\n",
      "process regulates process\n",
      "process positively regulates process\n",
      "process negatively regulates process\n",
      "affects expression of\n",
      "increases expression of\n",
      "decreases expression of\n",
      "affects uptake of\n",
      "decreases uptake of\n",
      "increases uptake of\n",
      "affects abundance of\n",
      "increases abundance of\n",
      "decreases abundance of\n",
      "affects expression in\n",
      "affects stability of\n",
      "decreases stability of\n",
      "increases stability of\n",
      "affects localization of\n",
      "decreases localization of\n",
      "increases localization of\n",
      "affects molecular modification of\n",
      "increases molecular modification of\n",
      "decreases molecular modification of\n",
      "affects activity of\n",
      "increases activity of\n",
      "decreases activity of\n",
      "affects degradation of\n",
      "increases degradation of\n",
      "decreases degradation of\n",
      "disrupts\n",
      "affects synthesis of\n",
      "increases synthesis of\n",
      "decreases synthesis of\n",
      "affects transport of\n",
      "decreases transport of\n",
      "increases transport of\n",
      "coexists with\n",
      "colocalizes with\n",
      "in complex with\n",
      "in cell population with\n",
      "in pathway with\n",
      "expresses\n",
      "translates to\n",
      "lacks part\n",
      "superclass of\n",
      "has molecular consequence\n",
      "in taxon\n",
      "derives from\n",
      "is metabolite of\n",
      "has phenotype\n",
      "in linkage disequilibrium with\n",
      "broad match\n",
      "gene product of\n",
      "affects risk for\n",
      "prevents\n",
      "predisposes\n",
      "participates in\n",
      "enables\n",
      "actively involved in\n",
      "capable of\n",
      "contributes to\n",
      "causes\n",
      "causes adverse event\n",
      "has completed\n",
      "related condition\n",
      "translation of\n",
      "phenotype of\n",
      "contributor\n",
      "editor\n",
      "author\n",
      "publisher\n",
      "provider\n",
      "treated by\n",
      "approved for treatment by\n",
      "produced by\n",
      "location of\n",
      "prevented by\n",
      "opposite of\n",
      "located in\n",
      "genetic association\n",
      "gene associated with condition\n",
      "condition associated with gene\n",
      "interacts with\n",
      "physically interacts with\n",
      "molecularly interacts with\n",
      "decreases molecular interaction\n",
      "increases molecular interaction\n",
      "directly interacts with\n",
      "genetically interacts with\n",
      "overlaps\n",
      "has part\n",
      "has excipient\n",
      "has food component\n",
      "has nutrient\n",
      "has variant part\n",
      "has active ingredient\n",
      "part of\n",
      "is active ingredient of\n",
      "food component of\n",
      "nutrient of\n",
      "is excipient of\n",
      "has gene product\n",
      "correlated with\n",
      "coexpressed with\n",
      "negatively correlated with\n",
      "has biomarker\n",
      "positively correlated with\n",
      "biomarker for\n",
      "affected by\n",
      "entity regulated by entity\n",
      "entity negatively regulated by entity\n",
      "entity positively regulated by entity\n",
      "process regulated by process\n",
      "process positively regulated by process\n",
      "process negatively regulated by process\n",
      "disrupted by\n",
      "occurs in\n",
      "develops from\n",
      "is sequence variant of\n",
      "is frameshift variant of\n",
      "is nonsense variant of\n",
      "is nearby variant of\n",
      "is non coding variant of\n",
      "is missense variant of\n",
      "is synonymous variant of\n",
      "is splice site variant of\n",
      "has increased amount\n",
      "has participant\n",
      "has output\n",
      "has input\n",
      "enabled by\n",
      "disease has basis in\n",
      "has sequence location\n",
      "derives into\n",
      "has metabolite\n",
      "subclass of\n",
      "has not completed\n",
      "expressed in\n",
      "produces\n",
      "has decreased amount\n",
      "temporally related to\n",
      "precedes\n",
      "preceded by\n",
      "similar to\n",
      "chemically similar to\n",
      "homologous to\n",
      "paralogous to\n",
      "xenologous to\n",
      "orthologous to\n",
      "model of\n",
      "transcribed from\n",
      "transcribed to\n",
      "manifestation of\n",
      "contraindicated for\n",
      "close match\n",
      "exact match\n",
      "same as\n"
     ]
    }
   ],
   "source": [
    "association_slots = tk.get_all_edge_properties(True)\n",
    "#pprint(association_slots)\n",
    "geno_class = tk.get_all_elements_by_mapping('GENO:0000536', True)\n",
    "#pprint(geno_class) \n",
    "all_classes = tk.get_all_classes()\n",
    "#pprint(all_classes)\n",
    "all_associations = tk.get_all_associations()\n",
    "predicate_descendents = tk.get_descendants(\"related to\")\n",
    "#geno_ancestors = tk.get_ancestors(\"Genotype\", False)\n",
    "predicate_map = {} \n",
    "mapping_list = []\n",
    "with open('/tmp/predicates_full.csv', 'w') as csv_file:\n",
    "    writer = csv.writer(csv_file)\n",
    "    for descendent in predicate_descendents:\n",
    "        p_element = tk.get_element(descendent)\n",
    "        for mapping in p_element.exact_mappings:\n",
    "            mapping_list.append(mapping)\n",
    "        for mapping_n in p_element.narrow_mappings:\n",
    "            mapping_list.append(mapping_n)\n",
    "        for mapping_b in p_element.broad_mappings:\n",
    "            mapping_list.append(mapping_b)\n",
    "        if 'RO' not in mapping_list:\n",
    "            #if tk.has_inverse(descendent):\n",
    "            #    print(descendent)\n",
    "            #    continue\n",
    "            #else:\n",
    "            for item in mapping_list:\n",
    "                #print(descendent + \",\" + item)\n",
    "                writer.writerow([descendent, mapping])\n",
    "            print(descendent)\n",
    "#with open('/tmp/predicates_full.csv', 'w') as csv_file:  \n",
    "#    writer = csv.writer(csv_file)\n",
    "#    for key, value in predicate_map.items():\n",
    "#       writer.writerow([key, value])\n",
    "    # pprint(p_element)\n",
    "#pprint(geno_descendents)\n",
    "#pprint(geno_ancestors)\n",
    "#pprint(all_associations)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fd2f97f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
