{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "64ff58e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:ToolkitGenerator:Range of slot 'treated by' (treatment) does not line with the domain of its inverse (treats)\n",
      "WARNING:ToolkitGenerator:Range of slot 'approved for treatment by' (treatment) does not line with the domain of its inverse (approved to treat)\n",
      "WARNING:ToolkitGenerator:Prefix case mismatch - supplied: NCBIGene expected: NCBIGENE\n",
      "WARNING:ToolkitGenerator:Prefix case mismatch - supplied: NCBIGene expected: NCBIGENE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "predicate does not descend from 'related_to: biolink:has_evidence\n",
      "missing predicate: biolink:enriched_in_genes_correlated_with\n",
      "missing predicate: biolink:transport_affected_by\n",
      "missing predicate: biolink:transport_affected_by\n"
     ]
    }
   ],
   "source": [
    "import bmt\n",
    "import requests\n",
    "import json\n",
    "from pprint import pprint\n",
    "\n",
    "tk = bmt.Toolkit()\n",
    "\n",
    "# /meta_knowledge_graph endpoint\n",
    "response = requests.get(\"https://tinyurl.com/ahn28nu5\")\n",
    "edges_and_nodes = response.json()\n",
    "missing_predicates = []\n",
    "mismatch_domain = []\n",
    "domains = []\n",
    "    \n",
    "for edge in edges_and_nodes.get('edges'):\n",
    "    subject = edge.get('subject')\n",
    "    predicate = edge.get('predicate')\n",
    "    p_object = edge.get('object')\n",
    "    if not tk.is_predicate(predicate):\n",
    "        # format the return as biolink:snake_case_string instead of snake_case_string\n",
    "        association_slots = tk.get_all_edge_properties(True)\n",
    "        if predicate in association_slots:\n",
    "            print(\"predicate does not descend from 'related_to: \" + predicate)\n",
    "        elif tk.is_mixin(predicate):\n",
    "                print(\"predicate is a mixin: \" + predicate)\n",
    "        else:\n",
    "            missing_predicates.append(predicate)\n",
    "            print(\"missing predicate: \" + predicate)\n",
    "    else:\n",
    "        domains = tk.get_ancestors(subject, True, True)\n",
    "        if subject not in domains:\n",
    "            print (\"domain mismatch: \" + tk.get_slot_domain(predicate))\n",
    "            mismatch_domain.append({\"predicate\": predicate,\n",
    "                                    \"domain_in_model\": tk.get_slot_domain(predicate), \n",
    "                                    \"domain_in_endpoint\": source})\n",
    "        if not tk.is_predicate(predicate):\n",
    "            association_slots = tk.get_all_edge_properties(True)\n",
    "            if predicate in association_slots:\n",
    "                print(\"predicate does not descend from 'related_to: \" + predicate)\n",
    "            elif tk.is_mixin(predicate):\n",
    "                print(\"predicate is a mixin: \" + predicate)\n",
    "            else:\n",
    "                missing_predicates.append(predicate)\n",
    "                print(\"missing predicate: \" + predicate)\n",
    "    \n",
    "        \n",
    "    \n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "9b32fef4",
   "metadata": {},
   "outputs": [
    {
     "ename": "AssertionError",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/2d/1xhwsfs15qb37bv4m62xn0tc0000gp/T/ipykernel_1035/2952314104.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32massert\u001b[0m \u001b[0mtk\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_mixin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'regulates'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32massert\u001b[0m \u001b[0mtk\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhas_inverse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'treats'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0;32massert\u001b[0m \u001b[0mtk\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhas_inverse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'close match'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAssertionError\u001b[0m: "
     ]
    }
   ],
   "source": [
    "assert tk.is_mixin('regulates')\n",
    "assert tk.has_inverse('treats')\n",
    "assert tk.has_inverse('close match')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "75fdb17d",
   "metadata": {},
   "outputs": [
    {
     "ename": "AssertionError",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/2d/1xhwsfs15qb37bv4m62xn0tc0000gp/T/ipykernel_1035/3419875488.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32massert\u001b[0m \u001b[0mtk\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_predicate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'process negatively regulates process'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0;32massert\u001b[0m \u001b[0mtk\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_predicate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'negatively regulates, process to process'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAssertionError\u001b[0m: "
     ]
    }
   ],
   "source": [
    "assert tk.is_predicate('process negatively regulates process')\n",
    "assert tk.is_predicate('negatively regulates, process to process')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "06fbb4c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import bmt\n",
    "import requests\n",
    "import json\n",
    "from pprint import pprint\n",
    "\n",
    "tk = bmt.Toolkit()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6a9cc2dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "related to\n",
      "related to at concept level\n",
      "related to at instance level\n",
      "associated with\n",
      "superclass of\n",
      "subclass of\n",
      "same as\n",
      "close match\n",
      "exact match\n",
      "broad match\n",
      "narrow match\n",
      "opposite of\n",
      "has real world evidence of association with\n",
      "active in\n",
      "acts upstream of\n",
      "has upstream actor\n",
      "acts upstream of positive effect\n",
      "has positive upstream actor\n",
      "acts upstream of negative effect\n",
      "has negative upstream actor\n",
      "acts upstream of or within\n",
      "has upstream or within actor\n",
      "acts upstream of or within positive effect\n",
      "has positive upstream or within actor\n",
      "acts upstream of or within negative effect\n",
      "has negative upstream or within actor\n",
      "mentions\n",
      "mentioned by\n",
      "contributor\n",
      "provider\n",
      "publisher\n",
      "editor\n",
      "author\n",
      "interacts with\n",
      "physically interacts with\n",
      "chemically interacts with\n",
      "molecularly interacts with\n",
      "genetically interacts with\n",
      "affects\n",
      "affected by\n",
      "associated with sensitivity to\n",
      "sensitivity associated with\n",
      "associated with resistance to\n",
      "resistance associated with\n",
      "diagnoses\n",
      "is diagnosed by\n",
      "increases amount or activity of\n",
      "decreases amount or activity of\n",
      "affects abundance of\n",
      "abundance affected by\n",
      "increases abundance of\n",
      "abundance increased by\n",
      "decreases abundance of\n",
      "abundance decreased by\n",
      "increases activity of\n",
      "affects activity of\n",
      "activity affected by\n",
      "activity increased by\n",
      "decreases activity of\n",
      "activity decreased by\n",
      "affects expression of\n",
      "expression affected by\n",
      "increases expression of\n",
      "expression increased by\n",
      "decreases expression of\n",
      "expression decreased by\n",
      "affects folding of\n",
      "folding affected by\n",
      "increases folding of\n",
      "folding increased by\n",
      "decreases folding of\n",
      "folding decreased by\n",
      "affects localization of\n",
      "localization affected by\n",
      "increases localization of\n",
      "localization increased by\n",
      "decreases localization of\n",
      "localization decreased by\n",
      "affects metabolic processing of\n",
      "metabolic processing affected by\n",
      "increases metabolic processing of\n",
      "metabolic processing increased by\n",
      "decreases metabolic processing of\n",
      "metabolic processing decreased by\n",
      "affects molecular modification of\n",
      "molecular modification affected by\n",
      "increases molecular modification of\n",
      "molecular modification increased by\n",
      "decreases molecular modification of\n",
      "molecular modification decreased by\n",
      "affects synthesis of\n",
      "synthesis affected by\n",
      "increases synthesis of\n",
      "synthesis increased by\n",
      "decreases synthesis of\n",
      "synthesis decreased by\n",
      "affects degradation of\n",
      "degradation affected by\n",
      "increases degradation of\n",
      "degradation increased by\n",
      "decreases degradation of\n",
      "degradation decreased by\n",
      "affects mutation rate of\n",
      "mutation rate affected by\n",
      "increases mutation rate of\n",
      "mutation rate increased by\n",
      "decreases mutation rate of\n",
      "mutation rate decreased by\n",
      "affects response to\n",
      "response affected by\n",
      "increases response to\n",
      "response increased by\n",
      "decreases response to\n",
      "response decreased by\n",
      "affects splicing of\n",
      "splicing affected by\n",
      "increases splicing of\n",
      "splicing increased by\n",
      "decreases splicing of\n",
      "splicing decreased by\n",
      "affects stability of\n",
      "stability affected by\n",
      "increases stability of\n",
      "stability increased by\n",
      "decreases stability of\n",
      "stability decreased by\n",
      "affects transport of\n",
      "transport affected by\n",
      "increases transport of\n",
      "transport increased by\n",
      "decreases transport of\n",
      "transport decreased by\n",
      "affects secretion of\n",
      "secretion affected by\n",
      "increases secretion of\n",
      "secretion increased by\n",
      "decreases secretion of\n",
      "secretion decreased by\n",
      "affects uptake of\n",
      "uptake affected by\n",
      "increases uptake of\n",
      "uptake increased by\n",
      "decreases uptake of\n",
      "uptake decreased by\n",
      "process regulates process\n",
      "process regulated by process\n",
      "process positively regulates process\n",
      "process positively regulated by process\n",
      "process negatively regulates process\n",
      "process negatively regulated by process\n",
      "entity regulates entity\n",
      "entity regulated by entity\n",
      "entity positively regulates entity\n",
      "entity positively regulated by entity\n",
      "entity negatively regulates entity\n",
      "entity negatively regulated by entity\n",
      "disrupts\n",
      "disrupted by\n",
      "gene product of\n",
      "has gene product\n",
      "transcribed to\n",
      "transcribed from\n",
      "translates to\n",
      "translation of\n",
      "homologous to\n",
      "paralogous to\n",
      "orthologous to\n",
      "xenologous to\n",
      "coexists with\n",
      "in pathway with\n",
      "in complex with\n",
      "in cell population with\n",
      "colocalizes with\n",
      "genetic association\n",
      "gene associated with condition\n",
      "condition associated with gene\n",
      "affects risk for\n",
      "risk affected by\n",
      "predisposes\n",
      "contributes to\n",
      "contribution from\n",
      "causes\n",
      "caused by\n",
      "ameliorates\n",
      "exacerbates\n",
      "treats\n",
      "treated by\n",
      "approved to treat\n",
      "approved for treatment by\n",
      "prevents\n",
      "prevented by\n",
      "correlated with\n",
      "positively correlated with\n",
      "negatively correlated with\n",
      "occurs together in literature with\n",
      "coexpressed with\n",
      "has biomarker\n",
      "biomarker for\n",
      "expressed in\n",
      "expresses\n",
      "has phenotype\n",
      "phenotype of\n",
      "occurs in\n",
      "contains process\n",
      "located in\n",
      "location of\n",
      "disease has location\n",
      "similar to\n",
      "chemically similar to\n",
      "has sequence location\n",
      "sequence location of\n",
      "model of\n",
      "models\n",
      "overlaps\n",
      "has part\n",
      "has plasma membrane part\n",
      "plasma membrane part of\n",
      "part of\n",
      "has input\n",
      "is input of\n",
      "has output\n",
      "is output of\n",
      "has participant\n",
      "catalyzes\n",
      "has catalyst\n",
      "has substrate\n",
      "is substrate of\n",
      "participates in\n",
      "actively involved in\n",
      "actively involves\n",
      "capable of\n",
      "capability of\n",
      "enables\n",
      "enabled by\n",
      "derives into\n",
      "derives from\n",
      "is metabolite of\n",
      "has metabolite\n",
      "food component of\n",
      "has food component\n",
      "nutrient of\n",
      "has nutrient\n",
      "is active ingredient of\n",
      "has active ingredient\n",
      "is excipient of\n",
      "has excipient\n",
      "manifestation of\n",
      "has manifestation\n",
      "produces\n",
      "produced by\n",
      "consumes\n",
      "consumed by\n",
      "temporally related to\n",
      "precedes\n",
      "preceded by\n",
      "directly interacts with\n",
      "affects expression in\n",
      "has variant part\n",
      "variant part of\n",
      "related condition\n",
      "is sequence variant of\n",
      "has sequence variant\n",
      "is missense variant of\n",
      "has missense variant\n",
      "is synonymous variant of\n",
      "has synonymous variant\n",
      "is nonsense variant of\n",
      "has nonsense variant\n",
      "is frameshift variant of\n",
      "has frameshift variant\n",
      "is splice site variant of\n",
      "has splice site variant\n",
      "is nearby variant of\n",
      "has nearby variant\n",
      "is non coding variant of\n",
      "has non coding variant\n",
      "disease has basis in\n",
      "causes adverse event\n",
      "adverse event caused by\n",
      "contraindicated for\n",
      "has contraindication\n",
      "has not completed\n",
      "not completed by\n",
      "has completed\n",
      "completed by\n",
      "decreases molecular interaction\n",
      "molecular interaction decreased by\n",
      "increases molecular interaction\n",
      "molecular interaction increased by\n",
      "in linkage disequilibrium with\n",
      "has increased amount\n",
      "increased amount of\n",
      "has decreased amount\n",
      "decreased amount in\n",
      "lacks part\n",
      "missing from\n",
      "develops from\n",
      "develops into\n",
      "in taxon\n",
      "taxon of\n",
      "has molecular consequence\n",
      "is molecular consequence of\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'NoneType' object has no attribute 'mixins'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Input \u001b[0;32mIn [3]\u001b[0m, in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m elements \u001b[38;5;241m=\u001b[39m tk\u001b[38;5;241m.\u001b[39mget_all_elements()\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m element \u001b[38;5;129;01min\u001b[39;00m elements:\n\u001b[0;32m----> 3\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[43mtk\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mis_predicate\u001b[49m\u001b[43m(\u001b[49m\u001b[43melement\u001b[49m\u001b[43m)\u001b[49m:\n\u001b[1;32m      4\u001b[0m         \u001b[38;5;28mprint\u001b[39m(element)\n",
      "File \u001b[0;32m~/Documents/src/bmt2/bmt/toolkit.py:863\u001b[0m, in \u001b[0;36mToolkit.is_predicate\u001b[0;34m(self, name, mixin)\u001b[0m\n\u001b[1;32m    844\u001b[0m \u001b[38;5;129m@lru_cache\u001b[39m(CACHE_SIZE)\n\u001b[1;32m    845\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mis_predicate\u001b[39m(\u001b[38;5;28mself\u001b[39m, name: \u001b[38;5;28mstr\u001b[39m, mixin: \u001b[38;5;28mbool\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28mbool\u001b[39m:\n\u001b[1;32m    846\u001b[0m     \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    847\u001b[0m \u001b[38;5;124;03m    Determines whether the given name is the name of an relation/predicate\u001b[39;00m\n\u001b[1;32m    848\u001b[0m \u001b[38;5;124;03m    in the Biolink Model. An element is a predicate if it descends from\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    861\u001b[0m \u001b[38;5;124;03m        That the named element is a valid relation/predicate in Biolink Model\u001b[39;00m\n\u001b[1;32m    862\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 863\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m RELATED_TO \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_ancestors\u001b[49m\u001b[43m(\u001b[49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmixin\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Documents/src/bmt2/bmt/toolkit.py:299\u001b[0m, in \u001b[0;36mToolkit.get_ancestors\u001b[0;34m(self, name, reflexive, formatted, mixin)\u001b[0m\n\u001b[1;32m    297\u001b[0m     ancs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mview\u001b[38;5;241m.\u001b[39mclass_ancestors(element\u001b[38;5;241m.\u001b[39mname, mixins\u001b[38;5;241m=\u001b[39mmixin, reflexive\u001b[38;5;241m=\u001b[39mreflexive)\n\u001b[1;32m    298\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(element, SlotDefinition):\n\u001b[0;32m--> 299\u001b[0m     ancs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mview\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mslot_ancestors\u001b[49m\u001b[43m(\u001b[49m\u001b[43melement\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmixins\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmixin\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreflexive\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mreflexive\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    300\u001b[0m     filtered_ancs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_filter_secondary(ancs)\n\u001b[1;32m    301\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/.virtualenvs/bmt2/lib/python3.8/site-packages/linkml_runtime/utils/schemaview.py:527\u001b[0m, in \u001b[0;36mSchemaView.slot_ancestors\u001b[0;34m(self, slot_name, imports, mixins, reflexive, is_a)\u001b[0m\n\u001b[1;32m    515\u001b[0m \u001b[38;5;129m@lru_cache\u001b[39m()\n\u001b[1;32m    516\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mslot_ancestors\u001b[39m(\u001b[38;5;28mself\u001b[39m, slot_name: SLOT_NAME, imports\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, mixins\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, reflexive\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, is_a\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m List[SlotDefinitionName]:\n\u001b[1;32m    517\u001b[0m     \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    518\u001b[0m \u001b[38;5;124;03m    Closure of slot_parents method\u001b[39;00m\n\u001b[1;32m    519\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    525\u001b[0m \u001b[38;5;124;03m    :return: ancestor slot names\u001b[39;00m\n\u001b[1;32m    526\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 527\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_closure\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43;01mlambda\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mx\u001b[49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mslot_parents\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mimports\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mimports\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmixins\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmixins\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mis_a\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mis_a\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mslot_name\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreflexive\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mreflexive\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.virtualenvs/bmt2/lib/python3.8/site-packages/linkml_runtime/utils/schemaview.py:45\u001b[0m, in \u001b[0;36m_closure\u001b[0;34m(f, x, reflexive, **kwargs)\u001b[0m\n\u001b[1;32m     43\u001b[0m i \u001b[38;5;241m=\u001b[39m todo\u001b[38;5;241m.\u001b[39mpop()\n\u001b[1;32m     44\u001b[0m visited\u001b[38;5;241m.\u001b[39mappend(i)\n\u001b[0;32m---> 45\u001b[0m vals \u001b[38;5;241m=\u001b[39m \u001b[43mf\u001b[49m\u001b[43m(\u001b[49m\u001b[43mi\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     46\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m v \u001b[38;5;129;01min\u001b[39;00m vals:\n\u001b[1;32m     47\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m v \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m visited:\n",
      "File \u001b[0;32m~/.virtualenvs/bmt2/lib/python3.8/site-packages/linkml_runtime/utils/schemaview.py:527\u001b[0m, in \u001b[0;36mSchemaView.slot_ancestors.<locals>.<lambda>\u001b[0;34m(x)\u001b[0m\n\u001b[1;32m    515\u001b[0m \u001b[38;5;129m@lru_cache\u001b[39m()\n\u001b[1;32m    516\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mslot_ancestors\u001b[39m(\u001b[38;5;28mself\u001b[39m, slot_name: SLOT_NAME, imports\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, mixins\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, reflexive\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, is_a\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m List[SlotDefinitionName]:\n\u001b[1;32m    517\u001b[0m     \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    518\u001b[0m \u001b[38;5;124;03m    Closure of slot_parents method\u001b[39;00m\n\u001b[1;32m    519\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    525\u001b[0m \u001b[38;5;124;03m    :return: ancestor slot names\u001b[39;00m\n\u001b[1;32m    526\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 527\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m _closure(\u001b[38;5;28;01mlambda\u001b[39;00m x: \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mslot_parents\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mimports\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mimports\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmixins\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmixins\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mis_a\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mis_a\u001b[49m\u001b[43m)\u001b[49m, slot_name, reflexive\u001b[38;5;241m=\u001b[39mreflexive)\n",
      "File \u001b[0;32m~/.virtualenvs/bmt2/lib/python3.8/site-packages/linkml_runtime/utils/schemaview.py:475\u001b[0m, in \u001b[0;36mSchemaView.slot_parents\u001b[0;34m(self, slot_name, imports, mixins, is_a)\u001b[0m\n\u001b[1;32m    468\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    469\u001b[0m \u001b[38;5;124;03m:param slot_name: child slot name\u001b[39;00m\n\u001b[1;32m    470\u001b[0m \u001b[38;5;124;03m:param imports: include import closure\u001b[39;00m\n\u001b[1;32m    471\u001b[0m \u001b[38;5;124;03m:param mixins: include mixins (default is True)\u001b[39;00m\n\u001b[1;32m    472\u001b[0m \u001b[38;5;124;03m:return: all direct parent slot names (is_a and mixins)\u001b[39;00m\n\u001b[1;32m    473\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    474\u001b[0m s \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mget_slot(slot_name, imports)\n\u001b[0;32m--> 475\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_parents\u001b[49m\u001b[43m(\u001b[49m\u001b[43ms\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mimports\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmixins\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mis_a\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.virtualenvs/bmt2/lib/python3.8/site-packages/linkml_runtime/utils/schemaview.py:448\u001b[0m, in \u001b[0;36mSchemaView._parents\u001b[0;34m(self, e, imports, mixins, is_a)\u001b[0m\n\u001b[1;32m    446\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_parents\u001b[39m(\u001b[38;5;28mself\u001b[39m, e: Element, imports\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, mixins\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, is_a\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m List[ElementName]:\n\u001b[1;32m    447\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m mixins:\n\u001b[0;32m--> 448\u001b[0m         parents \u001b[38;5;241m=\u001b[39m copy(\u001b[43me\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmixins\u001b[49m)\n\u001b[1;32m    449\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    450\u001b[0m         parents \u001b[38;5;241m=\u001b[39m []\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'NoneType' object has no attribute 'mixins'"
     ]
    }
   ],
   "source": [
    "elements = tk.get_all_elements()\n",
    "for element in elements:\n",
    "    if tk.is_predicate(element):\n",
    "        print(element)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff826e5b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
